File: typescript/exercises.ts
import { open } from "node:fs/promises";
import * as fs from 'fs';
import * as readline from 'readline';
import { constants } from "node:buffer";

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative");
  }
  let counts: Map<bigint, bigint> = new Map();
  let remaining = amount;
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination);
    remaining %= denomination;
  }
  return counts;
}

//From homework helper video
export function firstThenApply<T, U>(
  arr: T[],
  predicate: (x: T) => boolean,
  f: (x: T) => U): U | U[] | undefined {
  if (arr.length === 0) {
    return undefined;
  }
  const filteredArray = arr.find(predicate);
  if (filteredArray === undefined) {
    return undefined;
  }
  return f(filteredArray);
}

//From homework helper video
export function* powersGenerator(base: bigint): Generator<bigint> {
  let power = 0n;
  while (true) {
    yield base ** power;
    power++;
  }
}

//From homework helper video
export async function meaningfulLineCount(path: string): Promise<number> {
  let count = 0;
  const fileStream = fs.createReadStream(path);
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  });

  for await (const line of rl) {
    const trimmedLine = line.trim();
    if (trimmedLine.length > 0 && !trimmedLine.startsWith('#')) {
      count++;
    }
  }
  return count;
}

//From homework helper video
interface Sphere {
  kind: "Sphere";
  radius: number;
}

interface Box {
  kind: "Box";
  width: number;
  length: number;
  depth: number;
}

export type Shape = Sphere | Box;

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return 4 * Math.PI * Math.pow(shape.radius, 2);
    case "Box":
      return 2 * (shape.width * shape.length + shape.width * shape.depth + shape.length * shape.depth);
  }
}

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return (4 / 3) * Math.PI * Math.pow(shape.radius, 3);
    case "Box":
      return shape.width * shape.length * shape.depth;
  }
}

//From homework helper video
export interface BinarySearchTree<T> {
  size(): number;
  insert(value: T): BinarySearchTree<T>;
  contains(value: T): boolean;
  inorder(): Iterable<T>;
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0;
  }
  insert(value: T): BinarySearchTree<T> {
    return new Node(value, new Empty(), new Empty());
  }
  contains(_value: T): boolean {
    return false;
  }
  inorder(): Iterable<T> {
    return [];
  }
  toString(): string {
    return "()";
  }
}

class Node<T> implements BinarySearchTree<T> {
  constructor(
    private value: T,
    private left: BinarySearchTree<T>,
    private right: BinarySearchTree<T>
  ) { }

  size(): number {
    return 1 + this.left.size() + this.right.size();
  }

  insert(value: T): BinarySearchTree<T> {
    if (value < this.value) {
      return new Node(this.value, this.left.insert(value), this.right);
    } else if (value > this.value) {
      return new Node(this.value, this.left, this.right.insert(value));
    }
    return this;
  }

  contains(value: T): boolean {
    if (value === this.value) {
      return true;
    } else if (value < this.value) {
      return this.left.contains(value);
    } else {
      return this.right.contains(value);
    }
  }

  inorder(): Iterable<T> {
    function* inorderGenerator(node: Node<T>): IterableIterator<T> {
      if (node.left) yield* node.left.inorder();
      yield node.value;
      if (node.right) yield* node.right.inorder();
    }
    return inorderGenerator(this);
  }

  toString(): string {
    const leftStr = this.left.toString();
    const rightStr = this.right.toString();
    return `(${leftStr !== "()" ? leftStr : ""}${this.value}${rightStr !== "()" ? rightStr : ""})`;
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations


 (* Video homework helper *)
let first_then_apply array predicate consumer=
  match List.find_opt predicate array with 
  | None -> None
  | Some x -> consumer x


(* Video homework helper *)
let powers_generator base = 
   let rec generate_from power () = 
    Seq.Cons (power, generate_from (power * base))
  in 
  generate_from 1;;



    let meaningful_line_count filename =
      let meaningful_line line =
        let trimmed = String.trim line in
        String.length trimmed > 0 && not (String.get trimmed 0 = '#') (* String.starts_with instead *)
      in
      let the_file = open_in filename in
      let finally () = close_in the_file in
      let rec count_lines count =
        try
          let line = input_line the_file in
          if meaningful_line line then
            count_lines (count + 1)
          else
            count_lines count
        with
        | End_of_file -> count
      in
      Fun.protect ~finally (fun () -> count_lines 0)
    

 (* Video homework helper *)
 type shape =
 | Sphere of float
 | Box of float * float * float

let volume s =
 match s with
 | Sphere r -> Float.pi *. (r ** 3.) *. 4. /. 3.
 | Box (l, w, h) -> l *. w *. h

 let surface_area a =
  match a with
  | Sphere r -> Float.pi *. 4. *. r *. r
  | Box (l, w, h) -> 2.0 *. (l *. w +. w *. h +. h *. l)


 (* Video homework helper *)
 type 'a binary_search_tree =
 | Empty
 | Node of 'a binary_search_tree * 'a * 'a binary_search_tree

let rec size_helper = function
 | Empty -> 0
 | Node (l, _, r) -> 1 + size_helper l + size_helper r

let size tree = size_helper tree

let rec contains_helper v = function
 | Empty -> false
 | Node (l, x, r) -> v = x || contains_helper v (if v < x then l else r)

let contains v tree = contains_helper v tree

let rec inorder_helper = function
 | Empty -> []
 | Node (l, x, r) -> inorder_helper l @ [x] @ inorder_helper r

let inorder tree = inorder_helper tree

let rec insert_helper v = function
 | Empty -> Node (Empty, v, Empty)
 | Node (l, x, r) ->
   if v < x then Node (insert_helper v l, x, r)
   else if v > x then Node (l, x, insert_helper v r)
   else Node (l, x, r)

let insert v tree = insert_helper v tree
